lib/linguist/heuristics.rb

```diff
@@ -465,5 +465,13 @@ def call(data)
         Language["Scilab"]
       end
     end
+
+    disambiguate ".tsx" do |data|
+      if /^(import.+from|.+=\s*require)[\s\(]+['"]react['"]/.match(data)
+        Language["TypeScript"]
+      else
+        Language["XML"]
```

Comment:
I would have continued to the Bayesian classifier which classify files based on statistical properties.